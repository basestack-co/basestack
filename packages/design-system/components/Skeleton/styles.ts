import styled, { keyframes, css } from "styled-components";
import { rem } from "polished";

const pulsate = keyframes`
  0% {
    opacity: 1;
  }
  50% {
    opacity: 0.5;
  }
  100% {
    opacity: 1;
  }
`;

const flexColumn = css`
  display: flex;
  flex-direction: column;
`;

const flexRow = css`
  display: flex;
  flex-direction: row;
`;

const getValue = (value: number | string) =>
  typeof value === "string" ? value : rem(`${value}px`);

export const Container = styled.div<{
  backgroundColor?: string;
  padding?: number | string;
  hasShadow?: boolean;
  marginBottom: number;
}>`
  ${flexColumn};
  background-color: ${({ theme, backgroundColor }) =>
    backgroundColor || theme.skeleton.backgroundColor};
  border-radius: ${rem("4px")};
  padding: ${({ theme, padding }) => getValue(padding || theme.spacing.s5)};

  ${({ hasShadow, theme }) =>
    hasShadow &&
    css`
      box-shadow: ${theme.shadow.elevation2};
    `}

  &:not(:last-child) {
    margin-bottom: ${({ marginBottom }) => `${marginBottom}px`};
  }
`;

export const Wrapper = styled.div<{ displayInline?: boolean }>`
  ${({ displayInline }) => (displayInline ? flexRow : flexColumn)};
  animation: ${pulsate} 1.2s infinite linear;
`;

export const Item = styled.div<{
  height: number;
  width: number | string;
  marginBottom?: number | string;
  marginRight?: number | string;
  marginLeft?: number | string;
  isRound?: boolean;
}>`
  ${flexColumn};
  background-color: ${({ theme }) => theme.skeleton.color};
  height: ${({ height }) => rem(`${height}px`)};
  width: ${({ width }) => getValue(width || 0)};
  margin-bottom: ${({ marginBottom }) => getValue(marginBottom || 0)};
  margin-right: ${({ marginRight }) => getValue(marginRight || 0)};
  margin-left: ${({ marginLeft }) => getValue(marginLeft || 0)};
  border-radius: ${({ isRound }) => (isRound ? "50%" : rem("6px"))};
  flex-shrink: ${({ width }) =>
    typeof width === "string" && width.includes("%") ? "initial" : 0};
`;
