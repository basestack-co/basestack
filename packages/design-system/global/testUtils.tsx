import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
