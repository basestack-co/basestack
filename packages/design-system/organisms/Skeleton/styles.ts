import styled, { keyframes, css } from "styled-components";
import { rem } from "polished";

const pulsate = keyframes`
  0% {
    opacity: 1;
  }
  50% {
    opacity: 0.5;
  }
  100% {
    opacity: 1;
  }
`;

const flexColumn = css`
  display: flex;
  flex-direction: column;
`;

const flexRow = css`
  display: flex;
  flex-direction: row;
`;

export const Container = styled.div<{
  backgroundColor?: string;
  padding?: number;
  hasShadow?: boolean;
}>`
  ${flexColumn};
  background-color: ${({ theme, backgroundColor }) =>
    backgroundColor || theme.colors.white};
  border-radius: ${rem("4px")};
  padding: ${({ theme, padding }) =>
    padding ? rem(`${padding}px`) : theme.spacing.s5};

  ${({ hasShadow, theme }) =>
    hasShadow &&
    css`
      box-shadow: ${theme.shadow.elevation2};
    `}
`;

export const Wrapper = styled.div<{ displayInline?: boolean }>`
  ${({ displayInline }) => (displayInline ? flexRow : flexColumn)};
  animation: ${pulsate} 1.2s infinite linear;
`;

export const Item = styled.div<{
  height: number;
  width: number | string;
  marginBottom?: number;
  marginRight?: number;
}>`
  ${flexColumn};
  background-color: ${({ theme }) => theme.colors.gray100};
  border-radius: ${rem("6px")};
  height: ${({ height }) => rem(`${height}px`)};
  width: ${({ width }) =>
    typeof width === "string" ? width : rem(`${width}px`)};
  margin-bottom: ${({ marginBottom }) => rem(`${marginBottom}px`)};
  margin-right: ${({ marginRight }) => rem(`${marginRight}px`)};
`;
