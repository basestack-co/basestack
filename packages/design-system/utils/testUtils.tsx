import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
