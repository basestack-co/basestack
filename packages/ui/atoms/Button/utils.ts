import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
