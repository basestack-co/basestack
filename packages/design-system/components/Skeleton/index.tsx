import React, { memo } from "react";
// Components
import { Container, Item, Wrapper } from "./styles";

export interface SkeletonProps {
  numberOfItems?: number;
  gapBetweenItems?: number;
  backgroundColor?: string;
  padding?: number;
  displayInline?: boolean;
  hasShadow?: boolean;
  items: Array<{
    h: number;
    w: number | string;
    mb?: number | string;
    mr?: number | string;
    ml?: number | string;
    isRound?: boolean;
  }>;
}

// {Array.from(Array(3).keys()).map((item) => (

const Skeleton = ({
  numberOfItems = 1,
  gapBetweenItems = 0,
  backgroundColor,
  padding,
  items,
  hasShadow = true,
  displayInline = false,
}: SkeletonProps) => (
  <>
    {Array.from(Array(numberOfItems).keys()).map((item) => (
      <Container
        key={item}
        backgroundColor={backgroundColor}
        padding={padding}
        hasShadow={hasShadow}
        marginBottom={gapBetweenItems}
      >
        <Wrapper displayInline={displayInline}>
          {items.map((item, index) => (
            <Item
              key={index}
              height={item.h}
              width={item.w}
              marginBottom={item.mb || 0}
              marginRight={item.mr || 0}
              marginLeft={item.ml || 0}
              isRound={item.isRound}
            />
          ))}
        </Wrapper>
      </Container>
    ))}
  </>
);

export default memo(Skeleton);
