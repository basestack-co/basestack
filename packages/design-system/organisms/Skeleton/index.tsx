import React, { memo } from "react";
// Components
import { Container, Item, Wrapper } from "./styles";

interface SkeletonProps {
  backgroundColor?: string;
  padding?: number;
  displayInline?: boolean;
  hasShadow?: boolean;
  items: Array<{
    h: number;
    w: number | string;
    mb?: number;
    mr?: number;
  }>;
}

const Skeleton = ({
  backgroundColor,
  padding,
  items,
  hasShadow = true,
  displayInline = false,
}: SkeletonProps) => (
  <Container
    backgroundColor={backgroundColor}
    padding={padding}
    hasShadow={hasShadow}
  >
    <Wrapper displayInline={displayInline}>
      {items.map((item, index) => (
        <Item
          key={index}
          height={item.h}
          width={item.w}
          marginBottom={item.mb || 0}
          marginRight={item.mr || 0}
        />
      ))}
    </Wrapper>
  </Container>
);

export default memo(Skeleton);
