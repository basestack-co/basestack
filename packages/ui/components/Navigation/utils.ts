// Types
import { Translate } from "next-translate";
import { NextRouter } from "next/router";
// Utils
import { config } from "@basestack/utils";
import { signOut } from "next-auth/react";

export const getInternalLinks = (t: Translate, formId: string) => {
  return [
    {
      icon: "flag",
      activeText: "Submissions",
      to: `/form/${formId}/submissions`,
      text: t("internal.submissions"),
    },
    {
      icon: "flag",
      activeText: "Setup",
      to: `/form/${formId}/setup`,
      text: t("internal.setup"),
    },
    {
      icon: "settings",
      activeText: "Settings",
      to: `/form/${formId}/settings/general`,
      text: t("internal.settings"),
    },
  ];
};

export const getExternalLinks = (t: Translate) => {
  return [
    {
      icon: "description",
      activeText: "Documentation",
      to: config.urls.docs.flags.base,
      isExternal: true,
      text: t("external.docs"),
    },
    {
      icon: "link",
      activeText: "Github",
      to: config.urls.repo,
      isExternal: true,
      text: t("external.github"),
    },
  ];
};

export const getAvatarDropdownList = (
  t: Translate,
  router: NextRouter,
  onCreateForm: () => void,
) => {
  return [
    {
      icon: "add_circle",
      text: t("create.form"),
      onClick: onCreateForm,
      separator: true,
    },
    {
      icon: "settings",
      text: t("dropdown.settings"),
      onClick: () =>
        router.push({
          pathname: "/user/profile/settings",
        }),
    },
    {
      icon: "logout",
      text: t("dropdown.logout"),
      onClick: signOut,
    },
  ];
};
