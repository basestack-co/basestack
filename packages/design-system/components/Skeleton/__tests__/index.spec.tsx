import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import Skeleton from "..";

describe("Skeleton tests", () => {
  afterEach(cleanup);

  test("should render Skeleton correctly", () => {
    const { asFragment } = renderWithTheme(
      <Skeleton
        items={[
          { h: 25, w: "15%", mb: 10 },
          { h: 1, w: "100%", mb: 10 },
          { h: 50, w: "100%" },
        ]}
        padding={20}
        hasShadow
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
