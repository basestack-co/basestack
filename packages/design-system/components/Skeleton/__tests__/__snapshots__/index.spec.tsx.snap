// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton tests render Skeleton correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  flex-direction: column;
  background-color: #FFFFFF;
  border-radius: 0.25rem;
  padding: 1.25rem;
  box-shadow: rgba(0, 0, 0, 0.1) 0px 1px 3px 0px,rgba(0, 0, 0, 0.06) 0px 1px 2px 0px;
}

.c1 {
  display: flex;
  flex-direction: column;
  animation: jDLGYP 1.2s infinite linear;
}

.c2 {
  display: flex;
  flex-direction: column;
  background-color: #EEEEEE;
  border-radius: 0.375rem;
  height: 1.5625rem;
  width: 15%;
  margin-bottom: 0.625rem;
  margin-right: 0rem;
  margin-left: 0rem;
}

.c3 {
  display: flex;
  flex-direction: column;
  background-color: #EEEEEE;
  border-radius: 0.375rem;
  height: 0.0625rem;
  width: 100%;
  margin-bottom: 0.625rem;
  margin-right: 0rem;
  margin-left: 0rem;
}

.c4 {
  display: flex;
  flex-direction: column;
  background-color: #EEEEEE;
  border-radius: 0.375rem;
  height: 3.125rem;
  width: 100%;
  margin-bottom: 0rem;
  margin-right: 0rem;
  margin-left: 0rem;
}

<div
    class="c0"
    padding="20"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        height="25"
        marginbottom="10"
        marginleft="0"
        marginright="0"
        width="15%"
      />
      <div
        class="c3"
        height="1"
        marginbottom="10"
        marginleft="0"
        marginright="0"
        width="100%"
      />
      <div
        class="c4"
        height="50"
        marginbottom="0"
        marginleft="0"
        marginright="0"
        width="100%"
      />
    </div>
  </div>
</DocumentFragment>
`;
